File: ref/code/L11/DFS_FullPermutation.java
import java.io.*;
import java.util.*;

public class Dfs_FullPermutation {
	
	static int n;
	static int[] a, ans;
	static boolean[] vis;
	
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(in.readLine());
		a = new int[n+1];
		StringTokenizer st = new StringTokenizer(in.readLine());
	    for(int i=1; i<=n; i++) {
	        a[i] = Integer.parseInt(st.nextToken());
	    }
		ans = new int[n+1];
		vis = new boolean[1001];
		dfs(1);
	}

	static void dfs(int index) {
		if(index == n+1) {
			for(int i=1; i<=n; i++) {
				System.out.print(ans[i] + " ");	
			}
			System.out.println();
			return;
		}
		
		for(int i=1; i<=n; i++) {
			if(!vis[a[i]]) {
				ans[index] = a[i];
				vis[a[i]] = true;
				dfs(index+1);
				vis[a[i]] = false;			
			}
		}
		
	}
	
}

File: ref/code/L11/DFS_fullpermutation.cpp
#include <bits/stdc++.h>
using namespace std;

int a[10], cnt, n;
bool vis[1000];
int ans[10];

void dfs(int pos) {
    if (pos == n + 1) {
        cnt++;
        for (int i = 1; i <= n; i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
        return;
    }

    for (int i = 1; i <= n; i++) {
        if (!vis[a[i]]) {
            ans[pos] = a[i];
            vis[a[i]] = true;
            dfs(pos + 1);
            vis[a[i]] = false; // back trace
        }
    }
}

int main() {
    cin >> n;
    for(int i=1; i<=n; i++) {
        cin>>a[i];
    }
    dfs(1);
    cout << cnt << endl;
    return 0;
}